url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/UART.java

```diff
@@ -0,0 +1,346 @@
+            case REG_UARTIBRD:
+                //TODO: Not implemented
+                System.out.printf("UARTIBRD: 0x%08x\n", data);
+                break;
+            case REG_UARTFBRD:
+                //TODO: Not implemented
+                System.out.printf("UARTFBRD: 0x%08x\n", data);
+                break;
+            case REG_UARTLCR_H:
+                //TODO: Not implemented
+                System.out.printf("UARTLCR_H: 0x%08x\n", data);
+                super.writeWord(regaddr, data);
+                break;
+            case REG_UARTCR:
+                //TODO: Not implemented
+                //System.out.printf("UARTCR: 0x%08x\n", data);
+                super.writeWord(regaddr, data);
+                break;
+            case REG_UARTIFLS:
+                //TODO: Not implemented
+                System.out.printf("UARTIFLS: 0x%08x\n", data);
+                break;
+            case REG_UARTIMSC:
+                maskInt = data;
+                break;
+            case REG_UARTRIS:
+            case REG_UARTMIS:
+                //read only, ignored
+                break;
+            case REG_UARTICR:
+                rawInt &= ~data;
+                break;
+            case REG_UARTPeriphID0:
+            case REG_UARTPeriphID1:
+            case REG_UARTPeriphID2:
+            case REG_UARTPeriphID3:
+            case REG_UARTPCellID0:
+            case REG_UARTPCellID1:
+            case REG_UARTPCellID2:
+            case REG_UARTPCellID3:
+                //read only, ignored
+                break;
+            default:
+                super.writeWord(regaddr, data);
+                break;
+            }
+        }
+
+        @Override
+        public void run() {
+            mainLoop:
+            while (!shouldHalt()) {
+                try {
+                    while (strInput == null) {
+                        Thread.sleep(50);
+                        if (shouldHalt()) {
+                            break mainLoop;
+                        }
+                    }
+
+                    //NOTE: InputStream をポーリングします。
+                    //strInput が System.in かつ read() がブロックしたとき、
+                    //他のスレッドからブロッキングをキャンセルする方法がないため、
+                    //read() のブロックを避ける必要があるためです。
+                    //FIXME: この実装はマルチスレッドセーフではありません。
+                    //他スレッドが同時に strInput にアクセスする場合、
+                    //read() でブロックする可能性があります。
+                    while (strInput.available() == 0) {
+                        Thread.sleep(50);
+                        if (shouldHalt()) {
+                            break mainLoop;
+                        }
+                    }
+
+                    int c = strInput.read();
+                    if (c == -1) {
+                        //EOF
+                        break;
+                    }
+                    bufInput.append((char)c);
+
+                    intDst.setRaisedInterrupt(true);
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)